### PR TITLE
[sdk] Added support for setting `CardinalityLimit` allowed for the metric managed by the view.

### PR DESCRIPTION
**Pull request by Yun-Ting**
## Changes

Fixes https://github.com/open-telemetry/opentelemetry-dotnet/issues/5296
* Added support for setting CardinalityLimit allowed for the metric managed by the view.
* Defined OTEL1003 in docs in experimental API for setting Metric CardinalityLimit.
* Added .NET 8 ExperimentalAttribute decoration for setting Metric CardinalityLimit.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs [hunk 1]
```diff
@@ -91,11 +95,39 @@ public class MetricStreamConfiguration
...
+    /// will be used. The default value of 2000 will be used
+    /// if neither the view nor the MetricReader configures
+    /// MaxMetricPointsPerMetricStream.
+    /// </remarks>
+#if EXPOSE_EXPERIMENTAL_FEATURES
+    public
+#else
+    internal
+#endif
+    int? MaxMetricPointsPerMetricStream
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:116]: minor: `MetricStream` is already applied by the class name (`MetricStreamConfiguration.MaxMetricPointsPerMetricStream` is mouthful), consider one of the following names:
* `MetricStreamConfiguration.CardinalityLimit` (which is better IMHO)
* `MetricStreamConfiguration.MaxMetricPoints`
* `MetricStreamConfiguration.MaxAllowedMetricPoints`
* `MetricStreamConfiguration.MaxMetricPointsAllowed`

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs [hunk 2]
```diff
@@ -10,6 +12,8 @@ public class MetricStreamConfiguration
 {
     private string? name;
 
+    private int? maxMetricPointsPerMetricStream = 2000;
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:15]: I think either...

a) Set this to `null` initially.
or
b) Introduce a constant and set this and the sdk default to the value. So they don't get out of sync if one happens to change 😄

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs [hunk 3]
```diff
@@ -91,11 +95,39 @@ public class MetricStreamConfiguration
         }
     }
 
+    /// <summary>
+    /// Gets or sets a positive integer value
+    /// defining the maximum number of data points allowed to
+    /// per view.
+    /// </summary>
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:102]:
```suggestion
    /// <summary>
    /// Gets or sets a positive integer value
    /// defining the maximum number of data points allowed
    /// for the metric managed by the view.
    /// </summary>
```

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs [hunk 4]
```diff
@@ -91,11 +95,39 @@ public class MetricStreamConfiguration
...
+    /// <remarks>
+    /// Note: If there is no matching view, but the MetricReader
+    /// defines a default cardinality limit value based on the
+    /// instrument an aggregation is created for, that value
+    /// will be used. The default value of 2000 will be used
+    /// if neither the view nor the MetricReader configures
+    /// MaxMetricPointsPerMetricStream.
+    /// </remarks>
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:110]:
```suggestion
    /// <remarks>
    /// Note: If not set the SDK cardinality limit value will be used which
    /// defaults to 2000. Call <see
    /// cref="MeterProviderBuilderExtensions.SetMaxMetricPointsPerMetricStream"/>
    /// to configure the SDK defaults.
    /// </remarks>
```

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs [hunk 5]
```diff
@@ -91,11 +95,39 @@ public class MetricStreamConfiguration
...
+    internal
+#endif
+    int? MaxMetricPointsPerMetricStream
+    {
+        get => this.maxMetricPointsPerMetricStream;
+        set
+        {
+            if (value != null)
+            {
+                Guard.ThrowIfOutOfRange(value.Value, min: 1, max: int.MaxValue);
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:123]: i don't fully recollect Metrics design :( - but I think we do reserve 1 or 2 points for the zero tag and the overflow. should we ensure the minimum is 3 here to account for that?

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs [hunk 6]
```diff
@@ -91,11 +95,39 @@ public class MetricStreamConfiguration
         }
     }
 
+    /// <summary>
+    /// Gets or sets a positive integer value
+    /// defining the maximum number of data points allowed to
+    /// per view.
+    /// </summary>
+    /// <remarks>
+    /// Note: If there is no matching view, but the MetricReader
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:104]:
```suggestion
    /// Note: If there is no matching view, but the MeterProvider
```

I dont think we allow it per MetricReader, but this is rather applied to entire MeterProvider.. but please confirm..

Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:104]: It gets passed from the provider to the reader I think that's where the confusion is coming from. I suggested changing the remarks wholly above.

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:116]: While CardinalityLimit is better name, MaxMetricPoints ensures consistency across the MeterProvider setting and the View setting... but still, I am also thinking CardinalityLimit is better name and more aligned with spec.

(we added this feature much before spec, so we knew of the risk that spec will have different name than what we pick. 🤣 )

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs (hunk 6 above)
Comment (reyang) [reply src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:104]: This is got mixed up with the spec requirement https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#configuration-1 😄

1. A [view](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#view) with criteria matching the instrument an aggregation is created for has an aggregation_cardinality_limit value defined for the stream, that value SHOULD be used.
2. If there is no matching view, but the MetricReader defines a default cardinality limit value based on the instrument an aggregation is created for, that value SHOULD be used.
3. If none of the previous values are defined, the default value of 2000 SHOULD be used.

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs (hunk 5 above)
Comment (reyang) [reply src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:123]: Good topic for another PR 👍

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs [hunk 7]
```diff
@@ -91,11 +95,39 @@ public class MetricStreamConfiguration
...
+    /// defines a default cardinality limit value based on the
+    /// instrument an aggregation is created for, that value
+    /// will be used. The default value of 2000 will be used
+    /// if neither the view nor the MetricReader configures
+    /// MaxMetricPointsPerMetricStream.
+    /// </remarks>
+#if EXPOSE_EXPERIMENTAL_FEATURES
+    public
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:112]: @Yun-Ting As a follow-up PR let's add a diagnostic here. Something like this...

```csharp
#if NET8_0_OR_GREATER
    [Experimental(DiagnosticDefinitions.CardinalityLimitExperimentalApi, UrlFormat = DiagnosticDefinitions.ExperimentalApiUrlFormat)]
#endif
```

You can use this PR as a guide for how to do it: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5153

File: test/OpenTelemetry.Tests/Metrics/MetricViewTests.cs [hunk 8]
```diff
@@ -930,13 +959,18 @@ public void ViewConflict_TwoDistinctInstruments_ThreeStreams()
...
             {
                 return instrument.Description == "description1"
-                    ? new MetricStreamConfiguration() { Name = "MetricStreamB" }
-                    : new MetricStreamConfiguration() { Name = "MetricStreamC" };
+                    ? new MetricStreamConfiguration() { Name = "MetricStreamB", CardinalityLimit = 3 }
+                    : new MetricStreamConfiguration() { Name = "MetricStreamC", CardinalityLimit = 200000 };
+            })
+            .AddView((instrument) =>
+            {
+                // This view is ignored as the passed in MaxMetricPointsPerMetricStream is out of range.
```
Comment (reyang) [inline test/OpenTelemetry.Tests/Metrics/MetricViewTests.cs:972]:
```suggestion
                // This view is ignored as the passed in CardinalityLimit is out of range.
```

**Review by reyang — APPROVED** [review]

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs [hunk 9]
```diff
@@ -91,11 +95,36 @@ public class MetricStreamConfiguration
...
+    internal
+#endif
+    int? CardinalityLimit
+    {
+        get => this.cardinalityLimit;
+        set
+        {
+            if (value != null)
+            {
+                Guard.ThrowIfOutOfRange(value.Value, min: 3, max: int.MaxValue);
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:120]: @Yun-Ting would you put a comment here explaining why `min` is 3? (consider updating the XML doc to call out the accepted range)

Comment (Yun-Ting) [reply src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:120]: Yeah, one for the overflowAttribute when `emitOverflowAttribute` is set to true, and the other for 0 tag, assuming when users want set this value, they _want_ to passed in tags.

Currently, the lower limit for `.SetMaxMetricPointsPerMetricStream()` is [1](https://github.com/open-telemetry/opentelemetry-dotnet/blob/a31bca866dab2f9735713e9be9bcf80d95ca0104/src/OpenTelemetry/Metrics/Builder/MeterProviderBuilderExtensions.cs#L215C1-L215C59), though.

In [AggregatorStore](https://github.com/open-telemetry/opentelemetry-dotnet/blob/a31bca866dab2f9735713e9be9bcf80d95ca0104/src/OpenTelemetry/Metrics/AggregatorStore.cs#L96C9-L102C10), we reserved one for the overflowAttribute.

And in doc, it says one should be reserved for [0 tag](https://github.com/open-telemetry/opentelemetry-dotnet/tree/a31bca866dab2f9735713e9be9bcf80d95ca0104/docs/metrics/customizing-the-sdk#:~:text=One%20MetricPoint%20is%20reserved%20for%20every%20MetricStream%20for%20the%20special%20case%20where%20there%20is%20no%20key/value%20pair%20associated%20with%20the%20metric.%20The%20maximum%20number%20of%20MetricPoints%20has%20to%20accommodate%20for%20this%20special%20case.) (no key/value pair associated).

Also, there had been discussions on whether we should reserve 1 for zero dimension: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2635#discussion_r752490441.

Maybe this topic indeed worth another PR.

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs (hunk 1 above)
Comment (Yun-Ting) [reply src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs]: I've updated to use CardinalityLimit to align with the spec.

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs (hunk 7 above)
Comment (Yun-Ting) [reply src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:115]: Ended up using the same PR for adding the OTEL1003 definition/decoration.  :grin:
Hope this is not too nasty to review.

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs (hunk 5 above)
Comment (Yun-Ting) [reply src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:126]: Will use a follow-up PR to address this. 
Related discussion: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5312#discussion_r1478861696

File: src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs [hunk 10]
```diff
@@ -91,11 +98,39 @@ public class MetricStreamConfiguration
...
+#if EXPOSE_EXPERIMENTAL_FEATURES
+    /// <summary>
+    /// Gets or sets a positive integer value
+    /// defining the maximum number of data points allowed to for the metric
+    /// managed by the view.
+    /// </summary>
+    /// <remarks>
+    /// Note: If not set, the SDK cardinality limit value will be used, which
+    /// defaults to 2000. Call <see cref="MeterProviderBuilderExtensions"/>
+    /// SetMaxMetricPointsPerMetricStream"/> to confiture the SDK defaults.
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricStreamConfiguration.cs:110]:
```suggestion
    /// SetMaxMetricPointsPerMetricStream"/> to configure the MeterProvider defaults.
```

**Review by reyang — APPROVED** [review]
LGTM. The current test cases are using reflection, I think eventually we should remove reflection and test against the real limits (e.g. set a limit and see if we hit it or not). Not blocking this PR since it requires many changes in other places (including fixing some bugs).

**Pull request merged**